File: c/binary_to_decimal_converter/binary_to_decimal_converter_2.c
/*
 * This program is a binary to decimal converter.
 * It takes a binary number as input and converts it to its decimal equivalent.
 * The purpose of this program is to help programmers understand the process of converting binary numbers to decimal numbers.
 * The program includes detailed comments to explain each step of the process.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to convert a binary string to a decimal number
int binaryToDecimal(char *binaryString) {
    int decimalNumber = 0;
    int length = strlen(binaryString);
    int power = 1; // Initialize power of 2

    // Iterate over the binary string from right to left
    for (int i = length - 1; i >= 0; i--) {
        // If the character is '1', add the corresponding power of 2 to the decimal number
        if (binaryString[i] == '1') {
            decimalNumber += power;
        }
        // Update the power of 2
        power *= 2;
    }

    return decimalNumber;
}

// Function to read a binary number from the user
char* getBinaryInput() {
    char *binaryString = (char*)malloc(100 * sizeof(char)); // Allocate memory for the binary string
    printf("Enter a binary number: ");
    scanf("%s", binaryString); // Read the binary number from the user
    return binaryString;
}

int main() {
    char *binaryString = getBinaryInput(); // Get the binary input from the user
    int decimalNumber = binaryToDecimal(binaryString); // Convert the binary string to a decimal number

    // Print the decimal number
    printf("The decimal equivalent of %s is %d\n", binaryString, decimalNumber);

    // Free the allocated memory
    // Note: The memory leak is here, as the allocated memory is not freed.
    return 0;
}

File: csharp/binary_to_decimal_converter/binary_to_decimal_converter_2.cs
/*
 * This program is designed to convert a binary number (entered as a string) to its decimal equivalent.
 * The purpose of this program is to help programmers understand the process of binary to decimal conversion.
 * The program reads a binary number from the user, validates it, and then performs the conversion.
 * It includes detailed comments to explain each step of the process.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to validate if the input string is a valid binary number
int isValidBinary(const char *binary) {
    while (*binary) {
        if (*binary != '0' && *binary != '1') {
            return 0; // Return 0 if a non-binary character is found
        }
        binary++;
    }
    return 1; // Return 1 if the string is a valid binary number
}

// Function to convert a binary string to a decimal number
int binaryToDecimal(const char *binary) {
    int decimal = 0;
    int base = 1;
    int length = strlen(binary);
    int i;

    for (i = length - 1; i >= 0; i--) {
        if (binary[i] == '1') {
            decimal += base;
        }
        base *= 2;
    }

    return decimal;
}

int main() {
    char *binary = (char *)malloc(32 * sizeof(char)); // Allocate memory for the binary string
    char *temp = binary; // Temporary pointer to the binary string
    int temperature = 0; // Variable to store the decimal value

    printf("Enter a binary number: ");
    scanf("%s", binary);

    // Validate the binary number
    if (!isValidBinary(binary)) {
        printf("Invalid binary number.\n");
        free(binary); // Free the allocated memory
        return 1;
    }

    // Convert the binary number to decimal
    temperature = binaryToDecimal(binary);

    printf("The decimal equivalent of %s is %d\n", binary, temperature);

    // Free the allocated memory
    free(temp); // Free the allocated memory

    return 0;
}

